packages/expo-navigation-bar/android/src/main/java/expo/modules/navigationbar/NavigationBarModule.kt

```diff
@@ -1,201 +1,167 @@
+      }
     }
-  }
 
-  @ExpoMethod
-  fun setBehaviorAsync(behavior: String, promise: Promise) {
-    safeRunOnUiThread(promise) {
-      NavigationBar.setBehavior(it, behavior, { promise.resolve(null) }, { m -> promise.reject(ERROR_TAG, m) })
+    AsyncFunction("getBehaviorAsync") { promise: Promise ->
+      safeRunOnUiThread {
+        WindowInsetsControllerCompat(it.window, it.window.decorView).let { controller ->
+          val behavior = when (controller.systemBarsBehavior) {
+            // TODO: Maybe relative / absolute
+            WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE -> "overlay-swipe"
+            WindowInsetsControllerCompat.BEHAVIOR_SHOW_BARS_BY_SWIPE -> "inset-swipe"
+            // WindowInsetsControllerCompat.BEHAVIOR_SHOW_BARS_BY_TOUCH -> "inset-touch"
+            else -> "inset-touch"
+          }
+          promise.resolve(behavior)
+        }
+      }
     }
-  }
 
-  @ExpoMethod
-  fun getBehaviorAsync(promise: Promise) {
-    safeRunOnUiThread(promise) {
-      WindowInsetsControllerCompat(it.window, it.window.decorView).let { controller ->
-        val behavior = when (controller.systemBarsBehavior) {
-          // TODO: Maybe relative / absolute
-          WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE -> "overlay-swipe"
-          WindowInsetsControllerCompat.BEHAVIOR_SHOW_BARS_BY_SWIPE -> "inset-swipe"
-          // WindowInsetsControllerCompat.BEHAVIOR_SHOW_BARS_BY_TOUCH -> "inset-touch"
-          else -> "inset-touch"
+    OnStartObserving {
+      safeRunOnUiThread {
+        val decorView = it.window.decorView
+        @Suppress("DEPRECATION")
+        decorView.setOnSystemUiVisibilityChangeListener { visibility: Int ->
+          val isNavigationBarVisible = (visibility and View.SYSTEM_UI_FLAG_HIDE_NAVIGATION) == 0
+          val stringVisibility = if (isNavigationBarVisible) "visible" else "hidden"
+          sendEvent(
+            VISIBILITY_EVENT_NAME,
+            Bundle().apply {
+              putString("visibility", stringVisibility)
+              putInt("rawVisibility", visibility)
+            }
+          )
         }
-        promise.resolve(behavior)
       }
     }
-  }
 
-  /* Events */
-
-  @ExpoMethod
-  fun startObserving(promise: Promise) {
-    safeRunOnUiThread(promise) {
-      val decorView = it.window.decorView
-      @Suppress("DEPRECATION")
-      decorView.setOnSystemUiVisibilityChangeListener { visibility: Int ->
-        var isNavigationBarVisible = (visibility and View.SYSTEM_UI_FLAG_HIDE_NAVIGATION) == 0
-        var stringVisibility = if (isNavigationBarVisible) "visible" else "hidden"
-        mEventEmitter.emit(
-          VISIBILITY_EVENT_NAME,
-          Bundle().apply {
-            putString("visibility", stringVisibility)
-            putInt("rawVisibility", visibility)
-          }
-        )
+    OnStopObserving {
```

Comment:
```suggestion
    AsyncFunction("stopOberving") { promise: Promise ->
```
To access the promise, use `AsyncFunction` rather than `OnStopObserving`. The behavior is different because using `OnStopOberving` when switching threads will resolve the promise quicker, and JS won't wait for observer removal.